spec/message_spec.rb

```diff
@@ -0,0 +1,67 @@
+RSpec.describe "message encoding" do
+  # message Sally {
+  #   required string mood = 1;
+  #   optional uint32 age  = 2;
+  # }
+  class Sally
+    def initialize(attributes)
+      @attributes = {}
+      attributes.each do |attribute, value|
+        public_send("#{attribute}=", value)
+      end
+    end
+
+    ## Attributes
+
+    def age
+      @attributes[:age]
+    end
+
+    def age=(age)
+      raise ArgumentError, "age is not a number" unless age.respond_to?(:to_i)
+      raise ArgumentError, "age is too large" unless age.to_i <= 4_294_967_295
```

Comment:
In other words, give the user as much time as possible to cleanup/fixup their fields before they have to be right.

Since we only coerce/check when doing the actual encoding/decoding those are the only times those fields have to be correct.

My only push back on that idea, is that when we write a test, someone could write a test that said something like:
```ruby
let(:account){ ::Account.new(:balance => "your mother in law") }
```

And there would be no feedback to the user that the test case they just created is total crap because they couldn't possibly receive an account message with a string in the `balance` field.